### PR TITLE
feat: initial implementation for discriminator

### PR DESCRIPTION
**Pull request by marikaner**
This provides an initial implementation for the discriminating properties of OpenAPI specs: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/

According to the documentation this can only happen on `oneOf` and `anyOf` schemas and there can either be an explicit mapping or it is implicitly derived from the referenced schema names.

Of course, the Azure OpenAPI spec does not use `oneOf` but `type: object`. 

This PR got all the above covered. There is however, one more thing:

The Azure OpenAPI spec has a circular reference like:
```yml
- SchemaA:
  discriminator:
    propertyName: discriminatorProp
    mapping:
      value1: => SchemaB
      value2: => SchemaC
- SchemaB:
  allOf:
    - SchemaA // this references back to the original type
    - ...
```

This would result in something like this: 
```ts
type SchemaA = { discriminatorProp: 'value1'} & SchemaB;
type SchemaB = SchemaA & ... // <-- ERROR
```

Currently, the solution that I have in mind is to clean this up before even really parsing the schemas by removing such references.

### CODE REVIEW

File: packages/openapi-generator/src/parser/document.spec.ts [hunk 1]
```diff
@@ -103,6 +103,93 @@ describe('parseOpenApiDocument', () => {
     ]);
   });
 
+  it.only('parses discriminator schema with circular references', async () => {
```
Comment (deekshas8) [inline packages/openapi-generator/src/parser/document.spec.ts:106]:
```suggestion
  it('parses discriminator schema with circular references', async () => {
```

File: packages/openapi-generator/src/parser/schema.spec.ts [hunk 2]
```diff
@@ -518,18 +524,141 @@ describe('parseSchema', () => {
...
+            oneOf: [
+              { $ref: '#/components/schemas/SchemaA' },
+              { $ref: '#/components/schemas/SchemaB' }
+            ],
+            discriminator
+          },
+          await createTestRefs({
+            schemas: {
+              SchemaA: { properties: { c: { type: 'string' } } },
+              SchemaB: { properties: { c: { type: 'integer' } } }
```
Comment (MatKuhr) [inline packages/openapi-generator/src/parser/schema.spec.ts:561]: I might not fully understand this test, but I think it should be:

```suggestion
              SchemaA: { properties: { discriminatingProperty: { type: 'string' }, c: { type: 'string' } } },
              SchemaB: { properties: { discriminatingProperty: { type: 'string' }, c: { type: 'integer' } } }
```

The discriminator has to be a property of all schemas.

> It is important that all the models mentioned below anyOf or oneOf contain the property that the discriminator specifies.

Don't ask me why this redundancy is necessary, IDK 😄 
I guess it would also be possible for it to have different types, e.g. a string in `SchemaA` and a number in `SchemaB`.

Unfortunately, after checking again, the Azure spec seems to simply use this discriminator incorrectly. If you look at it in the swagger editor you can see that it is simply ignored and the type of the `messages` array is incorrect 😬

Comment (MatKuhr) [reply packages/openapi-generator/src/parser/schema.spec.ts:561]: Actually, they fixed this in the [preview 2024-08-01](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2024-08-01-preview/inference.json)

Comment (marikaner) [reply packages/openapi-generator/src/parser/schema.spec.ts:561]: Oh damn, good catch. I read this multiple times yesterday, but still ignored it ;) 
Unfortunately that makes things quite a bit more complicated, but let's see.

Comment (marikaner) [reply packages/openapi-generator/src/parser/schema.spec.ts:561]: ok, it doesn't make it more complicated. types should still work as expected.

**Comment by jjtang1985** [comment]
5 cases with `discriminator:` detected from the new [spec](https://github.com/SAP/ai-sdk-js/blob/generate-azure-openai-types/packages/foundation-models/src/azure-openai/spec/inference.yaml), where 2 of them are not relevant (`onYourData...`)

File: test-packages/test-services-openapi/test-service/schema/discriminator-object-entity-workaround-child-only-type.d.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+ * Copyright (c) 2024 SAP SE or an SAP affiliate company. All rights reserved.
+ *
+ * This is a generated file powered by the SAP Cloud SDK for JavaScript.
+ */
+import type { DiscriminatorObjectEntityWorkaroundParent } from './discriminator-object-entity-workaround-parent';
+/**
+ * Representation of the 'DiscriminatorObjectEntityWorkaroundChildOnlyType' schema.
+ */
+export type DiscriminatorObjectEntityWorkaroundChildOnlyType =
+  DiscriminatorObjectEntityWorkaroundParent;
```
Comment (deekshas8) [inline test-packages/test-services-openapi/test-service/schema/discriminator-object-entity-workaround-child-only-type.d.ts:11]: [q] This is still confusing to me, didn't we discuss that it should not reference the parent, but rather have properties of the parent?

**Review by deekshas8 — COMMENTED** [review]
Overall LGTM. I still have one question, maybe we can discuss offline.

File: .changeset/gorgeous-toys-kick.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@sap-cloud-sdk/openapi-generator': minor
+---
+
+[New Feature] Add support for discriminator properties in"one of" and "all of" schemas. Also add support for incorrect usage of the discriminator property with schemas of type "object".
```
[the remark this answers is not in the document]
Comment (marikaner) [reply .changeset/gorgeous-toys-kick.md:5]: We don't have a matching rule for this, but the rule for code font does not apply here I think (https://github.com/SAP/cloud-sdk/blob/main/STYLEGUIDE.md#use-code-font).

I will to rephrase it to make it apply better ;) 
```suggestion
[New Feature] Add support for discriminator properties in schemas that include `oneOf` and `anyOf`. Also add support for incorrect usage of the discriminator property with schemas of type `object`.
```

File: .changeset/gorgeous-toys-kick.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+'@sap-cloud-sdk/openapi-generator': minor
+---
+
+[New Feature] Add support for discriminator properties in schemas that include `oneOf` and `anyOf`. Also add support for incorrect usage of the discriminator property with schemas of type `object`.
```
Comment (marikaner) [inline .changeset/gorgeous-toys-kick.md:5]:
```suggestion
[New Functionality] Add support for discriminator properties in schemas that include `oneOf` and `anyOf`. Also add support for incorrect usage of the discriminator property with schemas of type `object`.
```

File: packages/openapi-generator/src/parser/document.ts [hunk 6]
```diff
@@ -24,16 +37,67 @@ export async function parseOpenApiDocument(
...
+    schemas
   };
 }
 
+type OpenApiPersistedSchemaWithDiscriminator = OpenApiPersistedSchema & {
+  schema: OpenApiOneOfSchema &
+    Required<Pick<OpenApiOneOfSchema, 'discriminator'>>;
+};
+
+async function sanitizeDiscriminatedSchemas(
```
Comment (marikaner) [inline packages/openapi-generator/src/parser/document.ts:57]:
```suggestion
// Some specs include incorrect discriminator definitions based on the schema type `object`, that circularly reference their parent type.
async function sanitizeDiscriminatedSchemas(
```

**Review by deekshas8 — APPROVED** [review]
LGTM.
